### DISCUSSION TITLE
Milvus build

### DISCUSSION THREAD

**Discussion by shaiktas** [milvus-io/milvus#30130] [Q&A and General discussion]
Hello All, We are trying to build the Milvus code base with our i internal build tools for CI/CD. Also we have plan to edit code and publish image with custom changes, so here we are looking some pointers how to build the milvus code base?

**Comment by yhmo** [comment]
https://github.com/milvus-io/milvus/blob/master/DEVELOPMENT.md

Make sure your local environment meets the prerequisites.
Basic steps:
```
# Clone github repository.
$ git clone https://github.com/milvus-io/milvus.git

# Install third-party dependencies.
$ cd milvus/
$ ./scripts/install_deps.sh

# Compile Milvus.
$ make
```

**Comment by shaiktas** [comment]
Thanks @yhmo for details, what is the expetion after execute make? Does it publish milvus image to artifactory? I executed locally it is taking very long time, but not sure what to verify. Can you please help here to understand what are verification steps after make? What the exact Docker file we should which published milvus image?

**Reply by yhmo** [reply]
"make" will download all the third-party libraries and compile the Milvus source code. 
![Screenshot from 2024-01-22 14-12-47](https://github.com/milvus-io/milvus/assets/2282099/0ac50161-2386-43d7-acc7-9a0fbe96056e)


The output is an executable binary under the "bin" folder:
![Screenshot from 2024-01-22 14-10-49](https://github.com/milvus-io/milvus/assets/2282099/985a299b-d5e5-476e-8cba-2fb84722286d)

After successfully compiling, you can run the two scripts under milvus repo:  ./scripts/start_standalone.sh **or** ./scripts/start_cluster.sh to start a local server.

**Reply by yhmo** [reply]
For making docker image, all the scripts are under this folder: https://github.com/milvus-io/milvus/tree/master/build
Read the README for more info.

An official pipeline to make milvus docker image is here: https://jenkins.milvus.io:18080/blue/organizations/jenkins/Milvus%20Publish%20Images/detail/2.3/305/pipeline/41/

You can learn from the pipeline.

**Comment by shaiktas** [comment]
Thanks @yhmo , after successful make exuection, i do see bin directory with milvus script.  when i tried started `scripts/start_standalone.sh` , it is failing, here attaching logs for reference, do i need to have first etcd, rocksMq running locally before i execute the `scripts/start_standalone.sh` ?? is this script start Milvus locally and can be accessible with localhost:19530?? Please help in clarifying . 
[standalone.log](https://github.com/milvus-io/milvus/files/14018295/standalone.log)

**Reply by yhmo** [reply]
Ok, if you want to run it locally, the full steps:
1. cmd to the project root path
2. source ./scripts/setenv.sh
3. use a text tool to open the config/milvus.yaml, replace all the "/var/lib/milvus" to be "/tmp/milvus"
4. cd deployments/docker/dev
5. docker-compose up -d
6. cd -
7. ./scripts/start_standalone.sh


Step 2 is to set environment path
Step 3 is to avoid the permission deny for some OS during milvus runtime
Step 4/5 is to start the third-party services, etcd/minio

**Reply by shaiktas** [reply]
Thank you, if my understanding is correct, Steps 4 and 5 to pull and start third party dependencies, step 7 will use locally compiled milvus code base (Which is result of make command),? after successful execution of `/scripts/start_standalone.sh` can i access milvus with localhost:19530??

**Reply by shaiktas** [reply]
@yhmo ^^

**Reply by yhmo** [reply]
If you use a text tool to open the scripts/start_standalone.sh, you will see it actually launches a local milvus server by the ./bin/milvus
The ./bin/milvus by default uses the ./config/milvus.yaml as configuration file.

Yes, localhost:19530 is the server address.

**Comment by shaiktas** [comment]
Hello @yhmo , thanks for the information, we are able to built milvus locally and also able to start server using /scipts/start_standalone.sh and changes made got relfected after checking standalone.log file. Now are are trying to set up CI/CD to publish image using /build/builder.sh script file, unfortunately our CI/CD framework does not allow us to use script file with docker compose commands. when i edited /build/builder.sh file it has below contents . So here are we using docker-compose builders to publish image?? is it possible to publish the image locally using thie builder.sh? IS it publishing image with locally compiled milvus from `/bin/milvus`??
```1
mkdir -p "${DOCKER_VOLUME_DIRECTORY:-.docker}/${IMAGE_ARCH}-${OS_NAME}-ccache"
mkdir -p "${DOCKER_VOLUME_DIRECTORY:-.docker}/${IMAGE_ARCH}-${OS_NAME}-go-mod"
mkdir -p "${DOCKER_VOLUME_DIRECTORY:-.docker}/${IMAGE_ARCH}-${OS_NAME}-vscode-extensions"
mkdir -p "${DOCKER_VOLUME_DIRECTORY:-.docker}/${IMAGE_ARCH}-${OS_NAME}-conan"
chmod -R 777 "${DOCKER_VOLUME_DIRECTORY:-.docker}"

docker-compose pull --ignore-pull-failures builder
docker-compose run --no-deps --rm builder "$@"

popd

------

**Reply by yhmo** [reply]
To make a local image:
1. make 
2. make install
3. ./build/build_image.sh


Step 1 is to compile the source code, output a binary milvus under the ./bin folder

Step 2 is to copy the dependent libraries into the ./lib folder
![Screenshot from 2024-01-25 14-33-32](https://github.com/milvus-io/milvus/assets/2282099/4ddb352f-0df8-44dd-b9cc-af094cad7d33)


Step 3 is to make a local docker image with tag "milvusdb/milvus:latest"
![Screenshot from 2024-01-25 14-40-36](https://github.com/milvus-io/milvus/assets/2282099/caa06337-1500-4238-8a21-d35c063786b3)

**Reply by shaiktas** [reply]
thanks @yhmo i am able to execute all instructions /commands on Mac successfully, but when we are integrating the mivlus code base with internal CI/CD Framework. We see some challenged, We are on Linux Platform. 

we are executing below scripts in this order:
      - source scripts/setenv.sh 
      - ./scripts/install_deps.sh
      - make
      - make install 

while executing `source scripts/setenv.sh `  we see
LIBJEMALLOC=/workspace/internal/core/output/lib/libjemalloc.so
22:23:38 -      ++ test -f /workspace/internal/core/output/lib/libjemalloc.so
22:23:38 -      ++ echo 'WARN: Cannot find /workspace/internal/core/output/lib/libjemalloc.so'
22:23:38 -      WARN: Cannot find /workspace/internal/core/output/lib/libjemalloc.so

we have below snipet in scripts/install_deps.sh.

----
sudo dnf install -y wget curl which \
      git make automake python3-devel \
      devtoolset-11-gcc devtoolset-11-gcc-c++ devtoolset-11-gcc-gfortran devtoolset-11-libatomic-devel \
      llvm-toolset-11.0-clang llvm-toolset-11.0-clang-tools-extra \
      libaio libuuid-devel zip unzip \
      ccache lcov libtool m4 autoconf automake

    sudo pip3 install conan==1.61.0
-------
Kidnly clarify on below
the first command is giving below 

Package epel-release-8-19.el8.noarch is already installed.
22:23:53 -      No match for argument: [1mcentos-release-scl-rh
22:23:53 -      Error: Unable to find a match: centos-release-scl-rh
Package wget-1.21.1-7.el9.x86_64 is already installed.
22:23:53 -      Package which-2.21-29.el9.x86_64 is already installed.
22:23:53 -      Package git-2.39.3-1.el9_2.x86_64 is already installed.
22:23:53 -      Package make-1:4.3-7.el9.x86_64 is already installed.
22:23:53 -      Package automake-1.16.2-8.el9.noarch is already installed.
22:23:53 -      No match for argument: [1mdevtoolset-11-gcc(B[m
22:23:53 -      No match for argument: [1mdevtoolset-11-gcc-c++(B[m
22:23:53 -      No match for argument: [1mdevtoolset-11-gcc-gfortran(B[m
22:23:54 -      No match for argument: [1mdevtoolset-11-libatomic-devel(B[m
22:23:54 -      No match for argument: [1mllvm-toolset-11.0-clang(B[m
22:23:54 -      No match for argument: [1mllvm-toolset-11.0-clang-tools-extra(B[m
22:23:54 -      Package zip-3.0-35.el9.x86_64 is already installed.
22:23:54 -      Package unzip-6.0-56.el9.x86_64 is already installed.
22:23:54 -      No match for argument: [1mccache(B[m
22:23:54 -      No match for argument: [1mlcov(B[m

I am wondering being with above , is all required packages mentioned in below command is got installed or not??

`sudo dnf install -y wget curl which \
      git make automake python3-devel \
      devtoolset-11-gcc devtoolset-11-gcc-c++ devtoolset-11-gcc-gfortran devtoolset-11-libatomic-devel \
      llvm-toolset-11.0-clang llvm-toolset-11.0-clang-tools-extra \
      libaio libuuid-devel zip unzip \
      ccache lcov libtool m4 autoconf automake`

as well when trying to execute `sudo pip3 install conan==1.61.0` 
getting pip3 not found ??
Appreciate your inputs to get this success on Linux .

**Reply by yhmo** [reply]
I have no idea why you get the error.
For pip3, you can install it: https://www.activestate.com/resources/quick-reads/how-to-install-and-use-pip3/

**Reply by shaiktas** [reply]
thanks @yhmo i am able to install pip3 using
dnf install -y python3 python3-pip
how about the below command?  any alternate for this? 

-----
yum install -y wget curl which \
      git make automake python3-devel \
      devtoolset-11-gcc devtoolset-11-gcc-c++ devtoolset-11-gcc-gfortran devtoolset-11-libatomic-devel \
      llvm-toolset-11.0-clang llvm-toolset-11.0-clang-tools-extra \
      libaio libuuid-devel zip unzip \
      ccache lcov libtool m4 autoconf automake
----------
i am getting below after this: How can we verify is above command executed successfully or errored out? 

-----Package epel-release-8-19.el8.noarch is already installed.
22:23:53 - No match for argument: �[1mcentos-release-scl-rh�
22:23:53 - Error: Unable to find a match: centos-release-scl-rh
Package wget-1.21.1-7.el9.x86_64 is already installed.
22:23:53 - Package which-2.21-29.el9.x86_64 is already installed.
22:23:53 - Package git-2.39.3-1.el9_2.x86_64 is already installed.
22:23:53 - Package make-1:4.3-7.el9.x86_64 is already installed.
22:23:53 - Package automake-1.16.2-8.el9.noarch is already installed.
22:23:53 - No match for argument: �[1mdevtoolset-11-gcc�(B�[m
22:23:53 - No match for argument: �[1mdevtoolset-11-gcc-c++�(B�[m
22:23:53 - No match for argument: �[1mdevtoolset-11-gcc-gfortran�(B�[m
22:23:54 - No match for argument: �[1mdevtoolset-11-libatomic-devel�(B�[m
22:23:54 - No match for argument: �[1mllvm-toolset-11.0-clang�(B�[m
22:23:54 - No match for argument: �[1mllvm-toolset-11.0-clang-tools-extra�(B�[m
22:23:54 - Package zip-3.0-35.el9.x86_64 is already installed.
22:23:54 - Package unzip-6.0-56.el9.x86_64 is already installed.
22:23:54 - No match for argument: �[1mccache�(B�[m
22:23:54 - No match for argument: �[1mlcov�(B�[m


---

**Reply by yhmo** [reply]
We didn't see this error before. You can search "Unable to find a match: centos-release-scl-rh" on google for more info.
